Add follow-up questions feature to AI chat

Muskan244 opened the pull request:
Closes #12243 
<!-- LINK THE ISSUE WITH THE "Closes" KEYWORD. Example: Closes https://github.com/JabRef/jabref/issues/13109 OR Closes #13109 -->

<!-- In about one to three sentences, describe the changes you have made: what, where, why, ... (REPLACE THIS LINE) -->
This PR implements a follow-up questions feature to AI chat. After the AI responds to a user's question, the system automatically generates 3-5 relevant follow-up questions that appear as clickable links below the chat. 
Changes:
-  Added new "Generate follow-up questions" setting in AI preferences
- Created `FollowUpQuestionGenerator` class to generate questions via AI
- Integrated follow-up generation into `AiChatLogic` after each AI response
- Added UI component to display clickable follow-up question links
- Questions are cleared when any new message is sent

<!-- NOTE: If your work is not yet complete, please open a draft pull request. In that case, outline your intended next steps. Do you need feedback? Will you continue in parallel? ... -->

### Steps to test

<!-- Describe how reviewers can test this fix/feature. Ideally, think of how you would guide a beginner user of JabRef to try out your change. -->
<!-- You can add screenshots or videos (using Loom - https://www.loom.com or by just adding .mp4 files). -->
<!-- (REPLACE THIS PARAGRAPH) -->
1.  Go to **Preferences → AI**, enable AI with API key, and check **"Generate follow-up questions"**
2. Open an entry with a PDF attached and click the **"AI chat"** tab
3. Send a question and verify that 3-5 clickable follow-up questions appear below the chat<img width="1470" height="837" alt="Screenshot 2025-11-17 at 2 21 31 PM" src="https://github.com/user-attachments/assets/67e35d43-4573-492e-9a21-36d45f203743" />
4. Click a follow-up question and verify it sends automatically and new questions appear
5. Disable the setting in preferences and verify no follow-up questions appear

<!-- YOU HAVE TO MODIFY THE ABOVE TEXT TO FIT YOUR PR. OTHERWISE, YOUR PR WILL BE CLOSED WITHOUT FURTHER COMMENT. -->

### Mandatory checks

<!--
Go through the checklist below. It is mandatory, even for a draft pull request.

Keep ALL the items. Replace the dots inside [.] and mark them as follows: 
[x] done 
[ ] TODO (yet to be done)
[/] not applicable
-->

- [x] I own the copyright of the code submitted and I license it under the [MIT license](https://github.com/JabRef/jabref/blob/main/LICENSE)
- [x] I manually tested my changes in running JabRef (always required)
- [/] I added JUnit tests for changes (if applicable)
- [x] I added screenshots in the PR description (if change is visible to the user)
- [x] I described the change in `CHANGELOG.md` in a way that is understandable for the average user (if change is visible to the user)
- [/] I checked the [user documentation](https://docs.jabref.org/): Is the information available and up to date? If not, I created an issue at <https://github.com/JabRef/user-documentation/issues> or, even better, I submitted a pull request updating file(s) in <https://github.com/JabRef/user-documentation/tree/main/en>.

InAnYan commented:
Okay, good! But the follow up questions template must be an `AiTemplate`

Muskan244 commented on CHANGELOG.md:
```diff
@@ -24,6 +24,7 @@ Note that this project **does not** adhere to [Semantic Versioning](https://semv
 - We added the possibility to configure the email provided to unpaywall. [#14340](https://github.com/JabRef/jabref/pull/14340)
 - We added a "Regenerate" button for the AI chat allowing the user to make the language model reformulate its response to the previous prompt. [#12191](https://github.com/JabRef/jabref/issues/12191)
 - We added support for transliteration of fields to English and automatic transliteration of generated citation key. [#11377](https://github.com/JabRef/jabref/issues/11377)
+- We added a "Generate follow-up questions" feature to AI chat that suggests relevant questions after each AI response. [#12243](https://github.com/JabRef/jabref/issues/12243)
```
Sure, I will update the `CHANGELOG.md`

koppor commented on jablib/src/main/java/org/jabref/logic/ai/AiDefaultPreferences.java:
```diff
@@ -58,6 +58,7 @@ public String toString() {
     public static final boolean ENABLE_CHAT = false;
     public static final boolean AUTO_GENERATE_EMBEDDINGS = false;
     public static final boolean AUTO_GENERATE_SUMMARIES = false;
+    public static final boolean GENERATE_FOLLOW_UP_QUESTIONS = false;
```
Options:

- nagging after some weeks: "Did you know?" -- I don't like this
- have it turned on - and enable quick turn off (we have this at user comment field)
- have it turned off - and have a quick-turn on possibility (as we do at Mr. DLib)

I would vote for option 3 - but as follow-up PR to have the PRs reviewable and get finished parts in main fast.

koppor commented on jabgui/src/main/java/org/jabref/gui/ai/components/aichat/AiChatComponent.java:
```diff
@@ -214,6 +217,39 @@ private void initializeChatPrompt() {
+        // Get the current questions (this is thread-safe)
+        List<String> questions = new ArrayList<>(aiChatLogic.getFollowUpQuestions());
+
+        // Update UI on the JavaFX Application Thread
```
```suggestion
```

koppor commented on jabgui/src/main/java/org/jabref/gui/ai/components/aichat/AiChatComponent.java:
```diff
@@ -214,6 +217,39 @@ private void initializeChatPrompt() {
+    }
+
+    private void updateFollowUpQuestions() {
+        // Get the current questions (this is thread-safe)
```
```suggestion
```

Muskan244 commented:
> * Isn't there a JavaFX element for increasing and decreasing numbers?

Yes, I think there is JavaFX `Spinner<Integer>`. I will replace text field so users can easily increase/decrease the number of follow-up questions.


> * Should we add "Automatically" at the new bullet point? - To be consistent to the other two settings?

I will also update it to include “Automatically” for consistency with the other bullet points.

Muskan244 commented:
I have implemented these changes:

- Replaced the number text field with a JavaFX Spinner
- Added tooltip on follow-up questions to view full text
- Updated wording to include “Automatically”, and chained bindings
- Replaced Platform.runLater with UiTaskExecutor.runInJavaFXThread
- Removed unnecessary comments

Let me know if anything else is needed!

Muskan244 commented on jabgui/src/main/java/org/jabref/gui/preferences/ai/AiTabViewModel.java:
```diff
@@ -37,13 +38,17 @@
 import de.saxsys.mvvmfx.utils.validation.Validator;
 
 public class AiTabViewModel implements PreferenceTabViewModel {
+    protected static SpinnerValueFactory<Integer> followUpQuestionsCountValueFactory = new SpinnerValueFactory.IntegerSpinnerValueFactory(1, 5, 3);
```
It is protected because it needs to be accessed from AiTab.java (a different class).

InAnYan commented:
> * [ ]  Isn't there a JavaFX element for increasing and decreasing numbers?
> 
>     * [ ]   Should we add "Automatically" at the new bullet point? - To be consistent to the other two settings?
> 
> 
> <img alt="grafik" width="451" height="195" src="https://private-user-images.githubusercontent.com/1366654/517595759-149d45cb-d1ee-482d-b8e0-3704c2e5cd0a.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NjM4OTc2MjIsIm5iZiI6MTc2Mzg5NzMyMiwicGF0aCI6Ii8xMzY2NjU0LzUxNzU5NTc1OS0xNDlkNDVjYi1kMWVlLTQ4MmQtYjhlMC0zNzA0YzJlNWNkMGEucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI1MTEyMyUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNTExMjNUMTEyODQyWiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9ZDQ5MzNkYzFjMzRhNWI4OGM3M2M0MWNlMjBmY2Q5NzNjY2I0YWY5YmZkNWJjMWFjMTZmYTY0OTNjYzU5OWI1NSZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.e5ILJ2HS-KhrKei33RWmAAvFoxpte3069GqxzydKCO8">

Replying to @koppor .

I think `Automatically` doesn't make sense. The reason why I introduced previous 2 options is because before one should've manually click "Generate summary" or "Embeddings".

Generation of follow-up questions already implies that they are generated automatically.

Though... maybe I was wrong with the wording "Automatically generate embeddings". Need to dive more

Muskan244 commented:
> And when there are follow-up questions, can you hide the `HBox` with example questions?

I just want to clarify whether this should be done in this PR or in a separate PR, because @koppor commented earlier:

> > > I think there are too many buttons for questions. What if we could hide "Try these examples" as soon as there are follow up questions?
> > 
> > 
> > I have not worked on this part yet. If you would like me to include it in this PR, please let me know and I will be happy to update it.
> 
> Very OK for me to do that in a separate PR.

If I am mixing up two different things, please let me know.

InAnYan commented:
Ah, sorry, now I understand your confusion.

But, IMHO, the review comments that are left are one-liners, so I think it's not hard to incorporate them in this PR

koppor commented:
> > > I have not worked on this part yet. If you would like me to include it in this PR, please let me know and I will be happy to update it.
> > Very OK for me to do that in a separate PR.
> If I am mixing up two different things, please let me know.

I mis-read your question. Please include it here.
